clippy_lints/src/empty_with_brackets.rs

```diff
@@ -97,38 +115,127 @@ impl EarlyLintPass for EmptyWithBrackets {
         }
     }
 
-    fn check_variant(&mut self, cx: &EarlyContext<'_>, variant: &Variant) {
+    fn check_variant(&mut self, cx: &LateContext<'_>, variant: &Variant<'_>) {
         let span_after_ident = variant.span.with_lo(variant.ident.span.hi());
 
-        if has_brackets(&variant.data) && has_no_fields(cx, &variant.data, span_after_ident) {
-            span_lint_and_then(
+        if has_no_fields(cx, Some(&variant.data), span_after_ident) {
+            match variant.data {
+                VariantData::Struct { .. } => {
+                    span_lint_and_then(
+                        cx,
+                        EMPTY_ENUM_VARIANTS_WITH_BRACKETS,
+                        span_after_ident,
+                        "enum variant has empty brackets",
+                        |diagnostic| {
+                            diagnostic.span_suggestion_hidden(
+                                span_after_ident,
+                                "remove the brackets",
+                                "",
+                                Applicability::MaybeIncorrect,
+                            );
+                        },
+                    );
+                },
+                VariantData::Tuple(.., local_def_id) => {
+                    // Don't lint reachable tuple enums
+                    if cx.effective_visibilities.is_reachable(variant.def_id) {
+                        return;
+                    }
+                    self.empty_tuple_enum_variants
+                        .entry(local_def_id)
+                        .or_insert(Usage::Unused {
+                            redundant_use_sites: vec![],
+                        });
+                },
+                VariantData::Unit(..) => {},
+            }
+        }
+    }
+
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let Some(def_id) = check_expr_for_enum_as_function(expr) {
+            if let Some(parentheses_span) = call_parentheses_span(cx.tcx, expr) {
+                match self.empty_tuple_enum_variants.get_mut(&def_id) {
+                    Some(Usage::Unused {
+                        ref mut redundant_use_sites,
+                    }) => {
+                        redundant_use_sites.push(parentheses_span);
+                    },
+                    None if has_no_fields(cx, None, parentheses_span) => {
```

Comment:
Why pass `None` here instead of the variant?

Reply:
Also, the span that's passed here is the wrong one. It expects the parentheses at the variant definition site, but this is passing the parentheses of the function call.

It doesn't explain the false positives, but passing this span makes `has_no_fields` check for the wrong thing (since it uses the span for checking for `#[cfg]`s).
TBH, I'd just get rid of the `empty_parentheses` function entirely and use [`!span_contains_cfg()`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/attrs/fn.span_contains_cfg.html) instead.